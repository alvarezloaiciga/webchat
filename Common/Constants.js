export const SupportedWebchatUrls = [
  'goquiq.com/app/webchat',
  'quiq-api.com/app/webchat',
  'quiq.sh/app/webchat',
  'centricient.com/app/webchat',
  'centricient.corp/app/webchat',
  'quiq.dev:7000/app/webchat',
  'quiq.dev:3000/app/webchat',
  'centricient.dev:7000/app/webchat',
  'quiq.dev:41014/app/webchat',
  'centricient.dev:41014/app/webchat',
];

export const unknownErrorMessage = 'An unknown error occurred.';

// Default launch button classes and id's
export const noAgentsAvailableClass = 'noAgentsAvailable';
export const mobileClass = 'mobile';
export const unsupportedBrowserClass = 'unsupportedBrowser';
export const storageDisabledClass = 'storageDisabled';
export const hasMobileNumberClass = 'hasMobileNumber';

export const chatClosedClass = 'ChatClosed';
export const chatOpenClass = 'ChatOpen';
export const quiqContainerId = 'quiqWebChat';

export const webchatPath = 'app/webchat/webchat.html';
export const bridgePath = 'app/webchat/bridge.html';

export const StandaloneWindowName = 'quiq-standalone-webchat';

export const eventTypes: {[string]: string} = {
  messageArrived: 'QUIQ_MESSAGE_ARRIVED',
  chatVisibilityDidChange: 'QUIQ_CHAT_VISIBILITY_DID_CHANGE',
  _standaloneOpen: 'QUIQ_STANDALONE_OPEN',
  _launchButtonVisibilityShouldChange: 'QUIQ_LAUNCH_BUTTON_VISIBILITY_SHOULD_CHANGE',
  _configurationDidChange: 'QUIQ_CONFIGURATION_DID_CHANGE',
  _chatIsReady: 'QUIQ_CHAT_IS_READY',
};

// Any eventType that does not begin with "_" is exposed on the Quiq object handed to the client.
export const publicEventTypes = Object.keys(eventTypes).reduce(
  (acc, k) => Object.assign({}, acc, k.startsWith('_') ? {} : {[k]: eventTypes[k]}),
  {},
);

export const MessageTypes = {
  TEXT: 'Text',
  ATTACHMENT: 'Attachment',
};

export const MessageStatus = {
  PENDING: 'pending',
  FAILED: 'failed',
};

export const Sentry = {
  endpoint:
    'https://sentry.io/api/170102/store/?sentry_key=5622397f17c44165a039d3b91d3e0193&sentry_version=7&sentry_client=Direct',
};

export const EventTypes = {
  JOIN: 'Join',
  LEAVE: 'Leave',
  REGISTER: 'Register',
  AGENT_TYPING: 'AgentTyping',
  AGENT_ENDED_CONVERSATION: 'AgentEndedConversation',
  END: 'End',
  SPAM: 'Spam',
  SEND_TRANSCRIPT: 'SendTranscript',
  ASSIGNED: 'Assigned',
};

export const AuthorTypes = {
  CUSTOMER: 'Customer',
  USER: 'User',
  SYSTEM: 'System',
};

export const AttachmentErrorTypes = {
  TOO_LARGE: 'attachmentTooLarge',
  UNSUPPORTED_TYPE: 'attachmentUnsupportedType',
  UPLOAD_ERROR: 'attachmentUploadError',
  EMPTY: 'emptyUpload',
};

// Define subsets of events
export const EndEventTypes = [EventTypes.END, EventTypes.SPAM];

export const DisplayElementTypes = [
  MessageTypes.TEXT,
  MessageTypes.ATTACHMENT,
  EventTypes.SEND_TRANSCRIPT,
  EventTypes.END,
  EventTypes.SPAM,
  EventTypes.ASSIGNED,
  AttachmentErrorTypes.TOO_LARGE,
  AttachmentErrorTypes.UNSUPPORTED_TYPE,
  AttachmentErrorTypes.UPLOAD_ERROR,
  AttachmentErrorTypes.EMPTY,
];

export const reduxActionTypes = {
  uploadProgress: 'UPLOAD_PROGRESS',
  addPendingMessage: 'ADD_PENDING_MESSAGE',
  updatePendingMessageId: 'UPDATE_PENDING_MESSAGE_ID',
  updatePlatformEvents: 'UPDATE_PLATFORM_EVENTS',
  configurationLoaded: 'CHAT_CONFIGURATION_LOADED',
  removeMessage: 'REMOVE_MESSAGE',
  isAgentAssigned: 'AGENT_ASSIGNED',
  addAttachmentError: 'ADD_ATTACHMENT_ERROR',
  markSendFailure: 'MARK_SEND_FAILURE',
};

export const postmasterActionTypes = {
  loadChat: 'QUIQ_LOAD_CHAT',
  unloadChat: 'QUIQ_UNLOAD_CHAT',
  getChatStatus: 'QUIQ_GET_CHAT_STATUS',
  setChatRegistrationField: 'QUIQ_SET_CHAT_REGISTRATION_FIELD',
  setChatVisibility: 'QUIQ_SET_CHAT_VISIBILITY',
  getChatVisibility: 'QUIQ_GET_CHAT_VISIBILITY',
  getAgentAvailability: 'QUIQ_GET_AGENT_AVAILABILITY',
  getHandle: 'QUIQ_GET_HANDLE',
  login: 'QUIQ_LOGIN',
  getCanFlashNotifications: 'QUIQ_GET_CAN_FLASH_NOTIFICATIONS',
  getMobileChatEnabled: 'QUIQ_GET_MOBILE_CHAT',
  getLocalStorage: 'QUIQ_GET_LOCAL_STORAGE',
};

export const ExtensionSdkEventTypes = {
  ESTIMATED_WAIT_TIME_CHANGED: 'estimatedWaitTimeChanged',
};

export const ChatInitializedState = {
  UNINITIALIZED: 'uninitialized',
  LOADING: 'loading',
  INITIALIZED: 'initialized',
  ERROR: 'error',
  DISCONNECTED: 'disconnected',
  BURNED: 'burned',
  SLEEPING: 'sleeping',
};

export const intlMessageTypes = {
  pageTitle: 'pageTitle',
  titleText: 'titleText',
  headerText: 'headerText',
  sendButtonLabel: 'sendButtonLabel',
  messageFieldPlaceholder: 'messageFieldPlaceholder',
  invalidEmailErrorMessage: 'invalidEmailErrorMessage',
  welcomeFormValidationErrorMessage: 'welcomeFormValidationErrorMessage',
  welcomeFormSubmitButtonLabel: 'welcomeFormSubmitButtonLabel',
  welcomeFormSubmittingButtonLabel: 'welcomeFormSubmittingButtonLabel',
  connectingMessage: 'connectingMessage',
  reconnectingMessage: 'reconnectingMessage',
  agentTypingMessage: 'agentTypingMessage',
  agentAssignedMessage: 'agentAssignedMessage',
  agentsNotAvailableMessage: 'agentsNotAvailableMessage',
  agentEndedConversationMessage: 'agentEndedConversationMessage',
  errorMessage: 'errorMessage',
  requiredFieldAriaLabel: 'requiredFieldAriaLabel',
  dockWindowTooltip: 'dockWindowTooltip',
  openInNewWindowTooltip: 'openInNewWindowTooltip',
  optionsMenuTooltip: 'optionsMenuTooltip',
  emailTranscriptMenuMessage: 'emailTranscriptMenuMessage',
  emailTranscriptMenuTooltip: 'emailTranscriptMenuTooltip',
  emailTranscriptInputPlaceholder: 'emailTranscriptInputPlaceholder',
  emailTranscriptInputCancelTooltip: 'emailTranscriptInputCancelTooltip',
  emailTranscriptInputSubmitTooltip: 'emailTranscriptInputSubmitTooltip',
  emailTranscriptInlineButton: 'emailTranscriptInlineButton',
  messageArrivedNotification: 'messageArrivedNotification',
  closeWindowTooltip: 'closeWindowTooltip',
  emojiPickerTooltip: 'emojiPickerTooltip',
  attachmentBtnTooltip: 'attachmentBtnTooltip',
  unsupportedFileType: 'unsupportedFileType',
  attachmentTooLarge: 'attachmentTooLarge',
  attachmentUploadError: 'attachmentUploadError',
  muteSounds: 'muteSounds',
  unmuteSounds: 'unmuteSounds',
  muteSoundsTooltip: 'muteSoundsTooltip',
  unmuteSoundsTooltip: 'unmuteSoundsTooltip',
  transcriptEmailedEventMessage: 'transcriptEmailedEventMessage',
  cannotStartNewConversationMessage: 'cannotStartNewConversationMessage',
  unsupportedOrientation: 'unsupportedOrientation',
  unableToSend: 'unableToSend',
  infectedFile: 'infectedFile',
  emptyUpload: 'emptyUpload',
};

export const localStorageKeys = ['quiq-data'];

export const MenuItemKeys = {
  EMAIL_TRANSCRIPT: 'emailTranscript',
  MUTE_SOUNDS: 'muteSounds',
};

export const UserEmailKey = 'quiq-client-data';

export const maxAttachmentSize = 25 * 1000000;

export const displayModes = {
  DOCKED: 'docked',
  UNDOCKED: 'undocked',
  EITHER: 'either',
};
