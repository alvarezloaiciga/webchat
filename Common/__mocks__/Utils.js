export const getBrowserName = () => 'Chrome';
export const getMajor = () => 60;
export const getDeviceType = () => 'PC';
export const getOSName = () => 'Windows';
export const isIPhone = () => false;
export const getUAInfo = () => ({
  browser: {
    name: 'Chrome',
    major: 60,
  },
  os: {
    name: 'Mac OS',
    version: 10,
  },
  device: {
    model: 'iPad',
    vendor: 'Apple',
    type: 'tablet',
  },
  cpu: {
    architecture: 'amd64',
  },
  engine: {
    name: 'WebKit',
    version: 537.36,
  },
  navigator:
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.81 Safari/537.36',
});
export const compatibilityMode = () => false;
export const isIE9 = () => false;
export const isIE10 = () => false;
export const isIOS = () => false;
export const nonCompatibleBrowser = () => false;
export const supportsFlexbox = () => true;
export const supportsSVG = () => true;
export const displayError = (error: string) => error;
export const inStandaloneMode = jest.fn(() => false);
export const isMobile = jest.fn(() => false);
export const isStorageEnabled = jest.fn(() => true);
export const isSupportedBrowser = jest.fn(() => true);
export const buildTemplateString = (s: string) => s;
export const enableEmailForCurrentConversation = jest.fn(() => true);
export const convertToExtensionMessages = jest.fn(() => []);
export const getHostingDomain = jest.fn(() => 'test.com');
export const domainIsAllowed = jest.fn(() => true);
