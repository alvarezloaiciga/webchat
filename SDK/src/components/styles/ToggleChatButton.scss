.ToggleChatButton {
  z-index: 99999;
  display: flex;
  justify-content: center;
  align-items: center;
  position: fixed;
  box-sizing: border-box;
  bottom: 24px;
  right: 24px;
  width: 60px;
  height: 60px;
  border: none;
  border-radius: 100%;
  color: #fff;
  fill: #fff;
  cursor: pointer;
  transition: 0.15s ease-in-out all;
  box-shadow: rgba(0, 0, 0, 0.117647) 0px 1px 6px, rgba(0, 0, 0, 0.117647) 0px 1px 4px;

  // Leave these in, some customer sites have default stylings for buttons that try to override these
  margin: 0; 
  padding: 0;

  &:hover {
    box-shadow: rgba(0, 0, 0, 0.156863) 0px 3px 10px, rgba(0, 0, 0, 0.227451) 0px 3px 10px;
  }

  &:focus {
    outline: none;
  }

  svg {
    fill: #fff;
    flex: 1 1 auto;
  }
}
