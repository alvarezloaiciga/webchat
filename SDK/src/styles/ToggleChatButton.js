export default {
  display: 'flex',
  justifyContent: 'center',
  alignItems: 'center',
  position: 'fixed',
  boxSizing: 'border-box',
  bottom: '24px',
  right: '24px',
  width: '60px',
  height: '60px',
  border: 'none',
  borderRadius: '100%',
  color: '#fff',
  fill: '#fff',
  cursor: 'pointer',
  transition: '0.15s ease-in-out all',
  boxShadow: 'rgba(0, 0, 0, 0.117647) 0px 1px 6px, rgba(0, 0, 0, 0.11 7647) 0px 1px 4px',
};
