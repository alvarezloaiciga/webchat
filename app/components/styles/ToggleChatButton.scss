#quiqWebChat {
  .ToggleChatButton {
    position: fixed;
    bottom: 24px;
    right: 24px;
    width: 60px;
    height: 60px;
    border: none;
    border-radius: 100%;
    color: #fff;
    cursor: pointer;
    transition: 0.15s ease-in-out all;
    box-shadow: rgba(0, 0, 0, 0.117647) 0px 1px 6px, rgba(0, 0, 0, 0.117647) 0px 1px 4px;

    &:hover {
      box-shadow: rgba(0, 0, 0, 0.156863) 0px 3px 10px, rgba(0, 0, 0, 0.227451) 0px 3px 10px;
    }

    &:focus {
      outline: none;
    }

    svg {
      fill: #fff;
    }
  }
}
