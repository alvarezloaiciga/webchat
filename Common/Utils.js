// @flow
declare var __DEV__: string;
declare var QuiqModernizr: Object;

import messages from 'Common/Messages';
import {SupportedWebchatUrls, localStorageKeys, unknownErrorMessage} from './Constants';
import {UAParser} from 'ua-parser-js';
import './modernizr';
import type {BrowserNames, DeviceTypes, OSNames, BrowserEngine, IntlMessage} from './types';

const parser = new UAParser();

export const getBrowserName = (): BrowserNames => parser.getResult().browser.name;

export const getEngineName = (): BrowserEngine => parser.getResult().engine.name;

export const getEngineVersion = (): number => parseInt(parser.getResult().engine.version, 10);

export const getMajor = (): number => parseInt(parser.getResult().browser.major, 10);

export const getDeviceType = (): DeviceTypes => parser.getResult().device.type;

export const getOSName = (): OSNames => parser.getResult().os.name;

export const getUAInfo = () => parser.getResult();

export const isMobile = () => !!getDeviceType();

export const isSupportedBrowser = () => {
  if (isMobile()) return true;

  const name = getBrowserName();
  const major = getMajor();

  if (name === 'Chrome' && major >= 43) return true;
  if (name === 'Firefox' && major >= 48) return true;
  if (name === 'Safari' && major >= 6.1) return true;
  if (name === 'Edge' && major >= 12) return true;
  if (name === 'IE' && major >= 10) return true;
  if (name === 'Opera' && major >= 13) return true;

  return false;
};

let storageEnabled;
export const isStorageEnabled = () => {
  if (typeof storageEnabled !== 'undefined') return storageEnabled;

  const storageKey = 'quiq-webchat-storage-test';
  const storageVal = 'enabled?';
  try {
    localStorage.setItem(storageKey, storageVal);
    if (localStorage.getItem(storageKey) !== storageVal) {
      storageEnabled = false;
    }
    localStorage.removeItem(storageKey);
    storageEnabled = true;
  } catch (e) {
    storageEnabled = false;
  }

  return storageEnabled;
};

export const compatibilityMode = () => {
  const compatList = [
    {
      name: 'IE',
      major: 10,
    },
  ];
  const {name, major} = parser.getResult().browser;
  return !!compatList.find(i => i.name === name && parseInt(major, 10) <= i.major);
};

export const getWebchatHostFromScriptTag = () => {
  // Determine host from the script tag that loaded webchat
  const scriptTags = Array.from(document.getElementsByTagName('script'));

  const mainScript = scriptTags.find(
    tag => tag.src && SupportedWebchatUrls.find(u => tag.src.toLowerCase().includes(u)),
  );

  if (!mainScript)
    return displayError('Unable to locate Quiq script tag for determining chat host.');

  const webchatUrl = mainScript.src;
  return webchatUrl.slice(0, webchatUrl.indexOf('/app/webchat'));
};

export const getWindowDomain = () => `${window.location.protocol}//${window.location.host}`;

export const isIFrame = (chatWindow: Object): boolean => {
  return chatWindow instanceof HTMLIFrameElement;
};

export const camelizeToplevelScreamingSnakeCaseKeys = (obj: Object) => {
  const newObject = {};
  Object.keys(obj).forEach(key => {
    const newKey = camelize(key.toLowerCase());
    newObject[newKey] = obj[key];
  });
  return newObject;
};

export const getHostingWindow = (): ?Object => {
  const parent = window.parent !== window.self ? window.parent : null;
  return window.opener || parent;
};

export const isIE9 = () => getBrowserName() === 'IE' && getMajor() <= 9;
export const isIE10 = () => getBrowserName() === 'IE' && getMajor() === 10;

export const nonCompatibleBrowser = () => getBrowserName() === 'IE' && getMajor() < 9;
// QuiqModernizr says IE10 doesn't support flexbox.
// It kind of does, at least for what we need it for... so go ahead and ignore QuiqModernizr in that case
export const supportsFlexbox = () => isIE10() || (QuiqModernizr.flexbox && QuiqModernizr.flexwrap);
export const supportsSVG = () =>
QuiqModernizr.svg && QuiqModernizr.svgfilters && QuiqModernizr.inlinesvg;

export const displayError = (error: string | IntlMessage, values: { [string]: string } = {}) => {
  const message = typeof error === 'string' ? error : (error.defaultMessage || unknownErrorMessage);
  throw new Error(buildTemplateString(message, values));
};

export const displayWarning = (error: string, values: { [string]: string } = {}) => {
  console.warn(buildTemplateString(error, values));
};

// If window.opener is not null, then we're in a popup.
export const inStandaloneMode = () => {
  // If this fails, it's because of a cross origin exception.
  // This means we  must be in an iFrame, since window and window.top are on different domains.
  // Thus, return false in the catch block.
  try {
    return window.self === window.top;
  } catch (e) {
    return false;
  }
};

export const getWebchatUrlFromScriptTag = () => {
  // Determine host from the script tag that loaded webchat
  const scriptTags = Array.from(document.getElementsByTagName('script'));

  const mainScript = scriptTags.find(
    tag => tag.src && SupportedWebchatUrls.find(u => tag.src.toLowerCase().includes(u)),
  );

  if (!mainScript) return displayError(messages.cannotFindScript);

  return mainScript.src;
};

export const camelize = (str: string) =>
  // First, lowercase all uppercase characters not adjacent to a lowercase character
  str
    .replace(/[A-Z]{2,}|^[A-Z]|[A-Z]$/g, a => a.toLowerCase())
    // Then remove symbols and uppercase starts of words
    .replace(/[_\- ]+([^_\- ])/g, (a, b) => b.toUpperCase());

export const inNonProductionCluster = () =>
  !!window.location.hostname.match(
    /.*[.-](dev|qa|perf)\.(centricient\.corp|quiq\.sh|goquiq\.com)/g,
  );

export const inLocalDevelopment = () =>
__DEV__ || !!window.location.hostname.match(/.*\.(centricient|quiq)\.dev/g);

/**
 * Retrieves all Quiq-related keys, including store.js metadata keys, from local storage.
 * @param contactPoint {string} - If defined, will search for keys within this CP namespace.
 * If undefined, will search for legacy, non-namespaced keys.
 * @param postfix {string} - If defined, will append this contact point as a namespace
 * to the end of each key in the return object.
 * Useful if we need to lookup non-namespaced keys, but want the returned object to be namespaced by contact point.
 * @returns {{}} - A map of local storage keys onto values.
 */
export const getQuiqKeysFromLocalStorage = (contactPoint: ?string, postfix: ?string): { [string]: any } => {
  try {
    if (!localStorage) return {};
    const ls = {};
    const cpPostfix = contactPoint ? `_${contactPoint}` : '';
    const allKeys = localStorageKeys.flatMap(k => [
      `${k}${cpPostfix}`,
      `__storejs_expire_mixin_${k}${cpPostfix}`,
      `__storejs_modified_timestamp_mixin_${k}${cpPostfix}`
    ]);

    allKeys.forEach(k => {
      const value = localStorage.getItem(k);
      if (value) {
        // If we a postfix is defined, add it to the key before we update the returned object.
        const modifiedKey = `${k}${postfix ? `_${postfix}` : ''}`;
        ls[modifiedKey] = value;
      }
    });

    return ls;
  } catch (e) {
    return {}; // localStorage Disabled. Pass it through until we can display
  }
};

export const setLocalStorageItemsIfNewer = (data: { [string]: any }) => {
  if (!localStorage) return;

  Object.keys(data).forEach(k => {
    // Ignore storejs metadata keys
    if (k.startsWith('__')) return;
    const newModifiedTime = data[`__storejs_modified_timestamp_mixin_${k}`] || 0;
    const existingModifiedTime = parseInt(
      localStorage.getItem(`__storejs_modified_timestamp_mixin_${k}`) || 0,
      10,
    );

    // Write a key to localStorage only if 1) the key does not currently exist or 2) the existing key was modified earlier than the new key
    if (!localStorage.getItem(k) || newModifiedTime > existingModifiedTime) {
      // Write key itself, plus storejs metadata keys
      localStorage.setItem(k, data[k]);
      if (data[`__storejs_modified_timestamp_mixin_${k}`]) {
        localStorage.setItem(
          `__storejs_modified_timestamp_mixin_${k}`,
          data[`__storejs_modified_timestamp_mixin_${k}`],
        );
      }
      if (data[`__storejs_expire_mixin_${k}`]) {
        localStorage.setItem(`__storejs_expire_mixin_${k}`, data[`__storejs_expire_mixin_${k}`]);
      }
    }
  });
};

export const clearQuiqKeysFromLocalStorage = () => {
  try {
    if (!localStorage) return;

    localStorageKeys.forEach(k => {
      localStorage.removeItem(k);
    });
  } catch (e) {} // eslint-disable-line no-empty
};

// From https://stackoverflow.com/questions/377961/efficient-javascript-string-replacement
export const buildTemplateString = (s: string, values: { [string]: any }): string => {
  return s.replace(/{(\w*)}/g, (m: string, key: string) => {
    return values.hasOwnProperty(key) ? values[key].toString() : '';
  });
};
