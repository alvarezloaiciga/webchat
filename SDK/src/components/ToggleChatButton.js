// @flow
/** @jsx h */
import {h} from 'preact';
import {getStyle} from 'Common/QuiqOptions';
import {getQuiqOptions} from 'Globals';
import {displayModes} from 'Common/Constants';
import InvitationBlurb from './InvitationBlurb';
import styled, {css} from 'preact-emotion';

export type ToggleChatButtonProps = {
  onClick: () => void,
  open: boolean,
  invitationBlurbText: string,
};

const StyledInvitationBubble = styled(InvitationBlurb)`
  position: fixed;
  right: 90px;
  bottom: 20px;
`;

const ToggleChatButtonStyle = css`
  z-index: 99999;
  display: flex;
  justify-content: center;
  align-items: center;
  position: fixed;
  box-sizing: border-box;
  bottom: 24px;
  right: 24px;
  width: 60px;
  height: 60px;
  border: none;
  border-radius: 100%;
  color: #fff;
  fill: #fff;
  cursor: pointer;
  transition: 0.15s ease-in-out all;
  box-shadow: rgba(0, 0, 0, 0.117647) 0px 1px 6px, rgba(0, 0, 0, 0.117647) 0px 1px 4px;

  // Leave these in, some customer sites have default stylings for buttons that try to override these
  margin: 0;
  padding: 0;

  &:hover {
    box-shadow: rgba(0, 0, 0, 0.156863) 0px 3px 10px, rgba(0, 0, 0, 0.227451) 0px 3px 10px;
  }

  &:focus {
    outline: none;
  }

  svg {
    fill: #fff;
    flex: 1 1 auto;
  }
`;

const ToggleChatButton = ({onClick, open, invitationBlurbText}: ToggleChatButtonProps) => {
  const {styles, colors, displayMode} = getQuiqOptions();

  return (
    <div>
      <button
        className={`ToggleChatButton ${ToggleChatButtonStyle}`}
        onClick={onClick}
        style={getStyle(styles.ToggleChatButton, {backgroundColor: colors.primary})}
      >
        <svg
          style={getStyle(styles.ToggleChatButtonIcon)}
          xmlns="http://www.w3.org/2000/svg"
          width="24"
          height="24"
          viewBox="0 0 24 24"
        >
          {open && displayMode !== displayModes.UNDOCKED ? (
            <path d="M23 20.168l-8.185-8.187 8.185-8.174-2.832-2.807-8.182 8.179-8.176-8.179-2.81 2.81 8.186 8.196-8.186 8.184 2.81 2.81 8.203-8.192 8.18 8.192z" />
          ) : (
            <path d="M12 1c-6.628 0-12 4.573-12 10.213 0 2.39.932 4.591 2.427 6.164l-2.427 5.623 7.563-2.26c9.495 2.598 16.437-3.251 16.437-9.527 0-5.64-5.372-10.213-12-10.213z" />
          )}
        </svg>
      </button>
      {invitationBlurbText && (
        <StyledInvitationBubble
          text={invitationBlurbText}
          style={getStyle(styles.InvitationBlurb, {
            backgroundColor: 'white',
            shadowColor: '#B2B2B2',
          })}
        />
      )}
    </div>
  );
};

export default ToggleChatButton;
