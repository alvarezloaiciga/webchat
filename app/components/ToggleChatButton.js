// @flow

import React from 'react';
import quiqOptions, {getStyle} from 'Common/QuiqOptions';
import ChatBubbleIcon from './ChatBubbleIcon';
import {connect} from 'react-redux';
import {isMobile} from 'Common/Utils';
import type {ChatState} from 'Common/types';
import './styles/ToggleChatButton.scss';

export type ToggleChatButtonProps = {
  toggleChat: () => void,
  chatContainerHidden: boolean,
};

const {colors, styles} = quiqOptions;

export const ToggleChatButton = ({toggleChat, chatContainerHidden}: ToggleChatButtonProps) => {
  const buttonStyle = getStyle(styles.ToggleChatButton, {backgroundColor: colors.primary});
  const iconStyle = getStyle(styles.ToggleChatButtonIcon);

  return (
    <button style={buttonStyle} onClick={toggleChat} className="ToggleChatButton">
      {!chatContainerHidden && !isMobile()
        ? <svg
            style={iconStyle}
            xmlns="http://www.w3.org/2000/svg"
            width="24"
            height="24"
            viewBox="0 0 24 24"
          >
            <path d="M23 20.168l-8.185-8.187 8.185-8.174-2.832-2.807-8.182 8.179-8.176-8.179-2.81 2.81 8.186 8.196-8.186 8.184 2.81 2.81 8.203-8.192 8.18 8.192z" />
            X
          </svg>
        : <ChatBubbleIcon style={iconStyle} />}
    </button>
  );
};

export default connect((state: ChatState) => ({
  chatContainerHidden: state.chatContainerHidden,
}))(ToggleChatButton);
