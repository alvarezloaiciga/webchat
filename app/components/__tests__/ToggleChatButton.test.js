// @flow
jest.mock('Common/Utils');
jest.mock('Common/QuiqOptions');

import React from 'react';
import type {ToggleChatButtonProps} from '../ToggleChatButton';
import {ToggleChatButton} from '../ToggleChatButton';
import {shallow} from 'enzyme';
import {isMobile} from 'Common/Utils';
import type {ShallowWrapper} from 'enzyme';

describe('ToggleChatButton component', () => {
  let wrapper: ShallowWrapper;
  let testProps: ToggleChatButtonProps;
  let render: () => void;

  beforeEach(() => {
    testProps = {
      toggleChat: jest.fn(),
      chatContainerHidden: true,
    };
    render = () => {
      wrapper = shallow(<ToggleChatButton {...testProps} />);
    };
  });

  describe('rendering', () => {
    it('renders', () => {
      render();
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('open', () => {
    it('switches svgs', () => {
      testProps.chatContainerHidden = false;
      render();

      expect(wrapper).toMatchSnapshot();
    });

    describe('mobile', () => {
      beforeEach(() => {
        (isMobile: any).mockReturnValueOnce(() => true);
      });

      it('always shows blurb sbg even when hidden is true', () => {
        render();
        expect(wrapper).toMatchSnapshot();
        (isMobile: any).mockReset();
      });
    });
  });
});
