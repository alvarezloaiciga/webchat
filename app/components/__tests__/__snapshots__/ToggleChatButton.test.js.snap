// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleChatButton component open mobile always shows blurb sbg even when hidden is true 1`] = `
<button
  className="ToggleChatButton"
  onClick={[Function]}
  style={
    Object {
      "backgroundColor": "#333",
    }
  }
>
  <ChatBubbleIcon
    style={Object {}}
  />
</button>
`;

exports[`ToggleChatButton component open switches svgs 1`] = `
<button
  className="ToggleChatButton"
  onClick={[Function]}
  style={
    Object {
      "backgroundColor": "#333",
    }
  }
>
  <svg
    height="24"
    style={Object {}}
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M23 20.168l-8.185-8.187 8.185-8.174-2.832-2.807-8.182 8.179-8.176-8.179-2.81 2.81 8.186 8.196-8.186 8.184 2.81 2.81 8.203-8.192 8.18 8.192z"
    />
    X
  </svg>
</button>
`;

exports[`ToggleChatButton component rendering renders 1`] = `
<button
  className="ToggleChatButton"
  onClick={[Function]}
  style={
    Object {
      "backgroundColor": "#333",
    }
  }
>
  <ChatBubbleIcon
    style={Object {}}
  />
</button>
`;
